### PR TITLE
feat: x-data and x-init alpine attribute support

### PR DESCRIPTION
**Pull request by dianfishekqi**
## Description
Hello @shufo , this PR adds support for formatting x-data and x-init attributes for those using alpine.

## Screenshot
image

There is more work to be done in order to preserve the attribute offset.

### CODE REVIEW

**Comment by dianfishekqi** [comment]
After preserving ident level
image

**Comment by dianfishekqi** [comment]
@shufo  I added a new test, and also tested it against a large codespace with a lot of cases, it would be great if you can take a look at the code and see if can spot any potential bugs.
This would improve the exp of a lot of laravel devs as currently incorrect indent causes issues with collapsing tags on editors.

**Comment by shufo** [comment]
@dianfishekqi 
Thank you very much for the PR! It's Great that you point that.

I fully agree with you about the x-data and x-init formats, but I am a little concerned about the format.
I know there are differing opinions here about the formatting, but I personally prefer to have the indent base position in the html tag or in the attribute like

### indentation based on html tag

```blade
@section('body')
    <div class="wrapper" x-data="{
        show: true,
        toggle() {
            this.show = !this.show;
        },
        get state() {
            return show ? 'open' : 'close';
        }
    }"
    x-init="() => alert('123')">
        <div id="app" x-data="{
            multi_line: @true,
            multi_line() {
                console.log();
            }
        }" class="wrapper">
            <div x-data="{ userId: @entangle('userId') }" id='div1'></div>

        </div>
    </div>
@stop

```

#### Example

https://github.com/edumarrom/dwec21/blob/7e5bcd40ce62a8dd9917d67d2e929aac65187c75/practicas/ud8/alpine/alpine-modulo.html#L35-L39

### indentation based on x-data attribute (I think it's better)

```blade
@section('body')
    <div class="wrapper" x-data="{
                             show: true,
                             toggle() {
                                 this.show = !this.show;
                             },
                             get state() {
                                 return show ? 'open' : 'close';
                             }
                         }"
        x-init="() => alert('123')">
        <div id="app" x-data="{
                          multi_line: @true,
                          multi_line() {
                              console.log();
                          }
                      }" class="wrapper">
            <div x-data="{ userId: @entangle('userId') }" id='div1'></div>

        </div>
    </div>
@stop
```

#### Example

https://github.com/rogue-agent/NewAlps/blob/6c45d33f2e89022066ffe07c10928ef2ac928071/src/app.html#L10-L15

It would be great if there are any conventions or practices in the Alpine.js community that we should follow regarding this format.

And I would like to hear any opinions you may have on this point.

**Comment by shufo** [comment]
Also, I think it is better to keep the format for inline x-data with a small number of elements like

```
x-data="{ hello: 'world' }
```

### Expected Behaviour

```html
x-data="{ hello: 'world' }
```

### Current behaviour

```
x-data="{
            hello: 'world'
        }"
```

It may be able to determine if it should be inline by counting the number of elements or by the indentation position.

**Comment by dianfishekqi** [comment]
@shufo 
Thanks for your suggestions, I do agree with preserving inline and the indentation based on the attribute looks great, I have added a new commit reflecting the changes.

**Comment by shufo** [comment]
@dianfishekqi 
Thanks for the fix!

But sorry I just realized that consistent indentation may be difficult to achieve with indentation based on x-data position because the position of the x-data can be in an odd column or in an even column.

```blade
    <div id="app" class="wrapper" x-data="{
                                    show: true,
```

```blade
        <div id="app" x-data="{
                        multi_line: @true,
```

I think it would be better to indent based on the position of the html tag, so that the indentation is consistent...how do you think?

```blade
@section('body')
    <div class="wrapper" x-data="{
        show: true,
        toggle() {
            this.show = !this.show;
        },
        get state() {
            return show ? 'open' : 'close';
        }
    }" x-init="() => {
        alert('123');
    }">
        <div id="app" x-data="{
            multi_line: @true,
            multi_line() {
                console.log();
            }
        }" class="wrapper">
            <div x-data="{ userId: @entangle('userId'), foo: 'bar', bar: 'baz', aaaaaaaaaaaaaa: 'bar' }" id='div1'></div>

        </div>
    </div>
@stop
```

The inlining fix was perfect. Thank you very much.

File: src/formatter.ts [hunk 1]
```diff
@@ -1148,6 +1188,47 @@ export default class Formatter {
...
   }
 
+  restoreXData(content: any) {
+    return _.replace(
+      content,
+      new RegExp(`^(.*?)${this.getXDataPlaceholder('(\\d+)')}`, 'gm'),
+      (_match: any, p1: any, p2:any) => {
+        const offsetLine = p1.split('\n').pop();
+        const matches  = (/(\s*)([^\s]+.*)/g).exec(offsetLine) ?? [];
+        const offset = (matches[1] ?? '').length + ((matches[2]??'').length / (this.indentCharacter === '\t' ? 4: 1));
```
Comment (shufo) [inline src/formatter.ts:1198]:
```suggestion
        const offset = (matches[1] ?? '').length / (this.indentCharacter === '\t' ? 4: 1));
```

File: src/formatter.ts [hunk 2]
```diff
@@ -1148,6 +1188,47 @@ export default class Formatter {
...
+        const lines = beautify.js_beautify(this.xData[p2],{brace_style:'preserve-inline'}).split('\n')
+
+        const firstLine = lines[0];
+        const indentLevel = Math.floor((offset - 2 * (this.indentCharacter === '\t' ? 1 : 4)) / this.indentSize);
```
Comment (shufo) [inline src/formatter.ts:1202]:
```suggestion
        const indentLevel = offset / this.indentSize;
```

File: src/formatter.ts [hunk 3]
```diff
@@ -1148,6 +1188,47 @@ export default class Formatter {
...
+  }
+
+  restoreXInit(content: any) {
+    return _.replace(
+      content,
+      new RegExp(`^(.*?)${this.getXInitPlaceholder('(\\d+)')}`, 'gm'),
+      (_match: any, p1: any, p2:any) => {
+        const offsetLine = p1.split('\n').pop();
+        const matches  = (/(\s*)([^\s]+.*)/g).exec(offsetLine) ?? [];
+        const offset = (matches[1] ?? '').length + ((matches[2]??'').length / (this.indentCharacter === '\t' ? 4: 1));
```
Comment (shufo) [inline src/formatter.ts:1218]:
```suggestion
        const offset = (matches[1] ?? '').length / (this.indentCharacter === '\t' ? 4: 1));
```

File: src/formatter.ts [hunk 4]
```diff
@@ -1148,6 +1188,47 @@ export default class Formatter {
...
+        
+        const lines = beautify.js_beautify(this.xInit[p2],{brace_style:'preserve-inline'}).split('\n')
+
+        const firstLine = lines[0];
+        const indentLevel = Math.floor((offset - 2 * (this.indentCharacter === '\t' ? 1 : 4)) / this.indentSize);
```
Comment (shufo) [inline src/formatter.ts:1223]:
```suggestion
        const indentLevel = offset / this.indentSize;
```

File: __tests__/fixtures/formatted.alpine_attributes.blade.php [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+                                    get state() {
+                                        return show ? 'open' : 'close';
+                                    }
+                                }" x-init="() => alert('123')">
+        <div id="app" x-data="{
+                        multi_line: @true,
+                        multi_line() {
+                            console.log()
+                        }
+                    }" class="wrapper">
```
Comment (shufo) [inline __tests__/fixtures/formatted.alpine_attributes.blade.php:16]:
```suggestion
        show: true,
        toggle() {
            this.show = !this.show;
        },
        get state() {
            return show ? 'open' : 'close';
        }
    }" x-init="() => alert('123')">
        <div id="app" x-data="{
            multi_line: @true,
            multi_line() {
                console.log();
            }
        }" class="wrapper">

```

**Review by shufo — COMMENTED** [review]
Added suggestion for implementation of indent based on html tag position.

**Comment by dianfishekqi** [comment]
Hello @shufo, switched indent to tag base, added a few scenarios  more to test

**Comment by shufo** [comment]
@dianfishekqi 
Thanks! Great!

**Comment by shufo** [comment]
Released at https://github.com/shufo/blade-formatter/releases/tag/v1.22.0

**Pull request merged**